deviceplugin/usb.go

```diff
@@ -0,0 +1,169 @@
+// A USB device must match exactly on all the given attributes to pass.
+type UsbSpec struct {
+	// Vendor is the USB Vendor ID of the device to match on.
+	// (Both of these get mangled to uint16 for processing - but you should use the hexadecimal representation.)
+	Vendor usbID `json:"vendor"`
+	// Product is the USB Product ID of the device to match on.
+	Product usbID `json:"product"`
+}
+
+type usbID uint16
+
+func (id *usbID) UnmarshalJSON(data []byte) error {
+	if string(data) == "null" || string(data) == `""` {
+		return nil
+	}
+	*id = usbID(binary.LittleEndian.Uint16(data))
+	return nil
+}
+
+func (id *usbID) String() string {
+	return fmt.Sprintf("%04x", int(*id))
+}
+
+// usbDevice represents a physical, tangible USB device.
+type usbDevice struct {
+	// Vendor is the USB Vendor ID of the device.
+	Vendor usbID `json:"vendor"`
+	// Product is the USB Product ID of the device.
+	Product usbID `json:"product"`
+	// Bus is the physical USB bus this device is located at.
+	Bus uint16 `json:"bus"`
+	// BusDevice is the location of the device on the Bus.
+	BusDevice uint16 `json:"busdev"`
+}
+
+func (dev *usbDevice) BusPath() (path string) {
+	return fmt.Sprintf(usbDevBus, dev.Bus, dev.BusDevice)
+}
+
+func enumerateUSBDevices() (specs []usbDevice, err error) {
+	allDevs, err := os.ReadDir(usbDevicesDir)
+	if err != nil {
+		return
+	}
+	for _, dev := range allDevs {
+		if !dev.IsDir() {
+			// There shouldn't be any raw files in this directory, but just in case.
+			continue
+		}
+		// Try to find the vendor ID file inside this device - this is a good indication that we're dealing with a device, not a bus.
+		vnd, err := os.ReadFile(dev.Name() + "/" + usbDevicesDirVendorIDFile)
+		if err != nil {
+			// We can't read the vendor file for some reason, it probably doesn't exist.
+			continue
+		}
+		prd, err := os.ReadFile(dev.Name() + "/" + usbDevicesDirProductIDFile)
+		if err != nil {
+			continue
+		}
+
+		// The following two calls shouldn't fail.
+		busRaw, err := os.ReadFile(dev.Name() + "/" + usbDevicesDirBusFile)
+		if err != nil {
+			return specs, err
+		}
+		bus := binary.LittleEndian.Uint16(busRaw)
+		busLocRaw, err := os.ReadFile(dev.Name() + "/" + usbDevicesDirBusDevFile)
+		if err != nil {
+			return specs, err
+		}
+		busLoc := binary.LittleEndian.Uint16(busLocRaw)
+
+		spec := usbDevice{
+			Vendor:    usbID(binary.LittleEndian.Uint16(vnd)),
+			Product:   usbID(binary.LittleEndian.Uint16(prd)),
+			Bus:       bus,
+			BusDevice: busLoc,
+		}
+		specs = append(specs, spec)
+	}
+	return
+}
+
+func searchUSBDevices(devices *[]usbDevice, vendor usbID, product usbID) (devs []usbDevice, err error) {
+	for _, dev := range *devices {
+		if dev.Vendor == vendor && dev.Product == product {
+			devs = append(devs, dev)
+		}
+	}
+	return
+}
+
+func (gp *GenericPlugin) discoverUSB() (devices []device, err error) {
+	for _, group := range gp.ds.Groups {
+		paths := make([]string, len(group.UsbSpecs))
```

Comment:
Here we are initializing a slice with both a length and a capacity equal to the length of `group.USBSpecs`, but later the things are appended to the paths slice, which means that we will end up with a bunch of empty paths before it.

Reply:
Thoughts on b10fc167?

Reply:
yes, that works too. Ideally, if we actually know the length ahead of time, then we would do
```go
paths := make([]string, 0, len(group.USBSpecs))
```
so that the underlying memory is allocated but the slice is length 0 and you can do `append`
but this is such a small optimization that it doesn't matter much

Reply:
I went ahead and reworked this whole block with concurrency, would appreciate a re-review 6cde66f

I _tend_ to prefer append() as it's a bit easier for the layperson to read, and as you said, it's such a small amount of memory that we're working with here that it basically doesn't really matter (unless someone comes along with 10,000 USB devices attached...)